Kernel/Storage/SD/SDHostController.cpp

```diff
@@ -0,0 +1,623 @@
+            if (issue_command(SD::CommandIndex::AppCmd, 0).is_error() || wait_for_response().is_error())
+                return false;
+
+            if (issue_command(SD::CommandIndex::AppSendOpCond, acmd41_arg)
+                    .is_error())
+                return false;
+
+            if (auto acmd41_response = wait_for_response();
+                !acmd41_response.is_error()) {
+
+                // 20. Check if the card supports the voltage windows we requested and SDHC
+                u32 response = acmd41_response.value().response[0];
+                if ((response & acmd41_voltage) != acmd41_voltage) {
+                    card_is_usable = false;
+                    return false;
+                }
+
+                ocr.raw = acmd41_response.value().response[0];
+            }
+
+            return ocr.card_power_up_status == 1;
+        })) {
+        return card_is_usable ? EIO : ENODEV;
+    }
+
+    // 6. If you requested to switch to 1.8V, and the card accepts, execute a voltage switch sequence
+    //    (we didn't ask it)
+
+    // 7. Send CMD2 (ALL_SEND_CID) to the card
+    TRY(issue_command(SD::CommandIndex::AllSendCid, 0));
+    auto all_send_cid_response = TRY(wait_for_response());
+    auto cid = bit_cast<SD::CardIdentificationRegister>(all_send_cid_response.response);
+
+    // 8. Send CMD3 (SEND_RELATIVE_ADDR) to the card
+    TRY(issue_command(SD::CommandIndex::SendRelativeAddr, 0));
+    auto send_relative_addr_response = TRY(wait_for_response());
+    u32 rca = send_relative_addr_response.response[0]; // FIXME: Might need to clear some bits here
+
+    // Extra steps:
+
+    TRY(issue_command(SD::CommandIndex::SendCsd, rca));
+    auto send_csd_response = TRY(wait_for_response());
+    auto csd = bit_cast<SD::CardSpecificDataRegister>(send_csd_response.response);
+
+    u32 block_count = (csd.device_size + 1) * (1 << (csd.device_size_multiplier + 2));
+    u32 block_size = (1 << csd.max_read_data_block_length);
+    u64 capacity = (u64)block_count * block_size;
+    u64 card_capacity_in_blocks = capacity / block_len;
+
+    // TODO: Do high speed initialisation, if supported
+    TRY(sd_clock_frequency_change(max_supported_sdsc_frequency));
+
+    TRY(issue_command(SD::CommandIndex::SelectCard, rca));
+    TRY(wait_for_response());
+
+    // Set block length to 512 if the card is SDSC.
+    // All other models only support 512 byte blocks so they don't need to be explicitly told
+    if (!ocr.card_capacity_status) {
+        TRY(issue_command(SD::CommandIndex::SetBlockLen, block_len));
+        TRY(wait_for_response());
+    }
+
+    auto scr = TRY(retrieve_sd_configuration_register(rca));
+
+    TRY(issue_command(SD::CommandIndex::AppCmd, rca));
+    TRY(wait_for_response());
+    TRY(issue_command(SD::CommandIndex::AppSetBusWidth, 0x2)); // 0b00=1 bit bus, 0b10=4 bit bus
+    TRY(wait_for_response());
+
+    return TRY(adopt_nonnull_lock_ref_or_enomem(
+        new SDMemoryCard(*this,
```

Comment:
again only `new (nothrow)` can actually return a nullptr signaling ENOMEM, so prefer that